### ISSUE TITLE
Performance regression of array copy on amd64

### ISSUE THREAD

**Issue by JamesKingdon** [eclipse/omr#4123]
This problem is being investigated in the context of the IBM SDK, so I apologize for the use of Java specific examples and version info that relates to that particular consumer of OMR.

The performance of StringBuilder.append() when appending small strings in the range of 1 to 20 characters is significantly worse than it was in earlier versions. Limited testing so far shows that performance was good in IBM 8 SR5 FP1 (OMR  - b033a01) and is bad in IBM 8 SR5 FP36 (OMR      - 06a046a). It has been suggested that a major part of the regression was introduced in FP7 but I have not confirmed this myself.

Initial profiling suggests that the problem may be related to the use of REP MOVSD to encode the copy, however there have been many changes in this area of the code and it is possible that other factors are involved.

A testcase is attached which demonstrates the problem by timing the creation of 10MByte strings by appending small strings of different lengths to an initially empty (but appropriately sized) StringBuilder.
[SBAppendTest.zip](https://github.com/eclipse/omr/files/3382520/SBAppendTest.zip)

Example results:

![StringBuilder append-perf](https://user-images.githubusercontent.com/26579150/61059760-8cf1ff80-a3c7-11e9-8879-331ce1bb8d61.jpg)

When appending single char strings, the new code is nearly 10x slower than the old code.

**Comment by JamesKingdon** [comment]
I'm currently looking at the history of omr/compiler/x/codegen/OMRTreeEvaluator.cpp to see if it is possible to restore the original encoding.

**Comment by fjeremic** [comment]
@JamesKingdon I think you may have meant to open this at eclipse/openj9? If so please close this issue and reopen there.

**Comment by JamesKingdon** [comment]
Isn't it more appropriate here if the code changes needed are in the OMR project?

**Comment by JamesKingdon** [comment]
OpenJ9 issue: https://github.com/eclipse/openj9/issues/6428

**Comment by mstoodle** [comment]
To be clear, I suggested that the issue be opened at OpenJ9 because it's a problem manifesting in an OpenJ9 scenario. The ultimate fix for that problem may very well reside in OMR, in which case it would make perfect sense to have this issue reopened and linked with the OpenJ9 one. It may only be easily verifiable via OpenJ9 though, so I recommended James start there.

Thanks for reporting the problem!
